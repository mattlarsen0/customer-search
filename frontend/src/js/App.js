import React, { useEffect, useState } from "react";
import { CustomerContext } from "./context/CustomerContext.js";
import { CustomerList } from "./components/CustomerList.js";
import { fetchCustomers } from "./utils/fetchCustomers.js";
import { CustomerSearchField } from "./components/CustomerSearchField.js";
import { CompanyDropdown } from "./components/CompanyDropdown.js";

export const App = () => {
    const [customerState, setCustomerState] = useState({
        customers: [],
        searchOptions: {
            name: "",
            companyName: ""
        },
        setSearchOptions: () => {},
        fetchAndSetCustomers: () => {},
    });
    
    useEffect(() => {
        customerState.setSearchOptions = (searchOptions) => {
            customerState.searchOptions = searchOptions;
            updateCustomerList(searchOptions, customerState, setCustomerState);

            // Update url to reflect new search options
            let urlParams = {};

            if (searchOptions.name)
            {
                urlParams.search = searchOptions.name;
            }

            if (searchOptions.companyName)
            {
                urlParams.filter_by_company_name = searchOptions.companyName;
            }

            var url = new URL("https://localhost:5001/");
            url.search = new URLSearchParams(urlParams).toString();

            window.history.pushState({}, "Customer Search Demo", url);
        };

        customerState.fetchAndSetCustomers = (searchOptions) => {
            updateCustomerList(searchOptions, customerState, setCustomerState);
        };
 
        updateCustomerList(customerState.searchOptions, customerState, setCustomerState);
    }, []);

    return (
        <CustomerContext.Provider value={customerState}>
            <div>
                <label>Search: <CustomerSearchField /></label>
            </div>
            <div>
                <label>Filter by company: <CompanyDropdown /></label>
            </div>
            <CustomerList />
        </CustomerContext.Provider>
    );
};

const updateCustomerList = async (searchOptions, customerState, setCustomerState) => {
    const result = await fetchCustomers(searchOptions);
    customerState.customers = result.customers;
    setCustomerState({
        ...customerState
    });
};