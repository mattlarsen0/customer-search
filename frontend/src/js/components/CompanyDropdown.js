import React from "react";
import { useCustomers } from "../hooks/useCustomers.js";
import { useCustomerSearchOptions } from "../hooks/useCustomerSearchOptions.js";

export const CompanyDropdown = () => {
    const {customers} = useCustomers();
    const {searchOptions, setSearchOptions} = useCustomerSearchOptions();

    var companies = [...new Set(customers.map(c => c.companyName))];

    var companyOptions = companies.map(c => (
        <option key={`company-${c}`} value={c}>{c}</option>
    ));

    const onChange = (event) => {
        searchOptions.companyName = event.target.value;
        setSearchOptions(searchOptions);
    };

    return (
        <select data-testid="select" onChange={onChange}>
            <option value="">None</option>
            {companyOptions}
        </select>
    );
};
