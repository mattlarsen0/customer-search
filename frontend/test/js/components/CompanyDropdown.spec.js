import { CompanyDropdown } from "../../../src/js/components/CompanyDropdown.js";
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

const mockSetSearchOptions = jest.fn();

jest.mock("../../../src/js/hooks/useCustomerSearchOptions.js", () => ({
    useCustomerSearchOptions: () => ({
        searchOptions: {
            name: "",
            companyName: ""
        },
        setSearchOptions: mockSetSearchOptions
    }),
    __esModule: true
}));

const testCompanyName = "test company";
const testCustomer1 = {
    id: 1,
    firstName: "first1",
    lastName: "last1",
    companyName: "test company"
};

const testCustomer2 = {
    id: 2,
    firstName: "first2",
    lastName: "last2",
    companyName: "company2"
};

const testCustomer3 = {
    id: 3,
    firstName: "first3",
    lastName: "last3",
    companyName: "company3"
};

jest.mock("../../../src/js/hooks/useCustomers.js", () => ({
    useCustomers: () => ({
        customers: [testCustomer1, testCustomer2, testCustomer3]
    }),
    __esModule: true
}));

test("can render", () => {
    expect(render(<CompanyDropdown />)).toMatchSnapshot();
});

test("does call hook on change", (done) => {
    render(<CompanyDropdown />);
    const companySelect = screen.getByTestId("select");

    fireEvent.change(companySelect, { target: { value: testCompanyName } });

    setTimeout(() => {
        expect(mockSetSearchOptions).toHaveBeenCalledWith({
            name: "",
            companyName: testCompanyName
        });
        expect(mockSetSearchOptions).toHaveBeenCalledTimes(1);
        done();
    }, 100);
});
