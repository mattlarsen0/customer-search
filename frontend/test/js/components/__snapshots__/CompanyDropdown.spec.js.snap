// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        data-testid="select"
      >
        <option
          value=""
        >
          None
        </option>
        <option
          value="test company"
        >
          test company
        </option>
        <option
          value="company2"
        >
          company2
        </option>
        <option
          value="company3"
        >
          company3
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      data-testid="select"
    >
      <option
        value=""
      >
        None
      </option>
      <option
        value="test company"
      >
        test company
      </option>
      <option
        value="company2"
      >
        company2
      </option>
      <option
        value="company3"
      >
        company3
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
